discovery_engine_core/core/src/stack/filters.rs

```diff
@@ -0,0 +1,121 @@
+// Copyright 2022 Xayn AG
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as
+// published by the Free Software Foundation, version 3.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <https://www.gnu.org/licenses/>.
+
+use crate::{document::Document, engine::GenericError};
+use url::Url;
+use xayn_discovery_engine_providers::Article;
+
+pub(crate) trait ArticleFilter {
+    fn apply(current: &[Document], articles: Vec<Article>) -> Result<Vec<Article>, GenericError>;
+}
+
+struct DuplicateFilter;
+impl ArticleFilter for DuplicateFilter {
+    fn apply(
+        current: &[Document],
+        mut articles: Vec<Article>,
+    ) -> Result<Vec<Article>, GenericError> {
+        let urls = current
+            .iter()
+            .map(|a| a.resource.url.as_str())
+            .collect::<Vec<_>>();
+        let filtered = articles
+            .drain(..)
+            .filter(|article| !urls.contains(&article.link.as_str()))
+            .collect();
+        Ok(filtered)
```

Comment:
instead of draining & filtering & collecting we could just retain the desired elements. similarly for the `MalformedImageUrlFilter`.
```suggestion
        articles.retain(|article| !urls.contains(&article.link.as_str()));

        Ok(articles)
```